src/main/groovy/org/transmartproject/batch/biodata/BioExperimentDAO.groovy

```diff
@@ -0,0 +1,94 @@
+package org.transmartproject.batch.biodata
+
+import com.google.common.collect.ImmutableSet
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.dao.EmptyResultDataAccessException
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert
+import org.springframework.stereotype.Component
+import org.transmartproject.batch.clinical.db.objects.Sequences
+import org.transmartproject.batch.clinical.db.objects.Tables
+import org.transmartproject.batch.db.SequenceReserver
+
+/**
+ * Retrieval and insertion of bio experiments.
+ */
+@Component
+@Slf4j
+@CompileStatic
+class BioExperimentDAO {
+    @Value(Tables.BIO_EXPERIMENT)
+    private SimpleJdbcInsert bioExperimentInsert
+
+    @Autowired
+    private NamedParameterJdbcTemplate jdbcTemplate
+
+    @Autowired
+    private SequenceReserver sequenceReserver
+
+    private final Set<String> validKeys = ImmutableSet.of(
+            'bio_experiment_type',
+            'etl_id',
+            'title',)
+
+    Long findOrCreateBioExperiment(String studyId,
+                                   Map<String, Object> extraData = [:]) {
+        assert studyId != null
+        def intersection = extraData.keySet().intersect(validKeys)
+        if (intersection) {
+            throw new IllegalArgumentException("Invalid keys: $intersection")
+        }
+
+        Map<String, Object> desiredValues = [
+                accession: (Object) studyId,
+                etl_id: "METADATA:$studyId" as String,
+                title: 'Metadata not available',
+        ]
+        desiredValues.putAll(extraData)
+        log.debug("Desired values for bio experiment are: $desiredValues")
+
+        try {
+            jdbcTemplate.update("LOCK TABLE $Tables.BIO_EXPERIMENT " +
```

Comment:
Actually I don't need anymore. A GWAS dataset has several analyses (which may or many not share the same "study name") that are inserted simultaneously and in an earlier version of my change set, I was creating the `bio_experiment` row in those separate threads, so I needed locking. But now I insert all the bio_experiments in the same step.

In any case, it may be better this way, because it increases re-usability of the `DAO`.